pkg/cloudprovider/overlay/cloudprovider.go

```diff
@@ -0,0 +1,197 @@
+		if err != nil {
+			return []*cloudprovider.InstanceType{}, err
+		}
+		removeNodePoolRequirements(nodePool, its)
+	}
+	return its, nil
+}
+
+func OverlayInstanceTypes(ctx context.Context, kubeClient client.Client, instanceTypes []*cloudprovider.InstanceType) ([]*cloudprovider.InstanceType, error) {
+	overlays := &v1alpha1.NodeOverlayList{}
+	err := kubeClient.List(ctx, overlays)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+
+	// Only apply overlays that have passed runtime validation and know there have not been
+	// any updates to the spec of the overlay field
+	validOverlays := lo.Filter(overlays.Items, func(o v1alpha1.NodeOverlay, _ int) bool {
+		condition := o.StatusConditions().Get(v1alpha1.ConditionTypeValidationSucceeded)
+		return condition.IsTrue() && condition.ObservedGeneration == o.Generation
+	})
+
+	// We order the overlays from largest to smallest, as we have already done validation
+	// we can be sure that we will not be blocked due to invalid or conflicting overlays
+	sort.Slice(validOverlays, func(x int, y int) bool {
+		return lo.FromPtr(validOverlays[x].Spec.Weight) > lo.FromPtr(validOverlays[y].Spec.Weight)
+	})
+	work := []func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType){
+		overlayCapacityOnInstanceTypes,
+		overlayPriceOnInstanceTypes,
+	}
+	lop.ForEach(work, func(f func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType), i int) {
+		f(validOverlays, instanceTypes)
+	})
+
+	return instanceTypes, nil
+}
+
+func overlayCapacityOnInstanceTypes(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType) {
+	for _, overlay := range overlays {
+		overlaySelector := scheduling.NewRequirements()
+		overlaySelector.Add(scheduling.NewNodeSelectorRequirements(overlay.Spec.Requirements...).Values()...)
+
+		for _, it := range its {
+			if it.Requirements.IsCompatible(overlaySelector) {
+				it.Capacity = lo.Assign(it.Capacity, overlay.Spec.Capacity)
+				it.ApplyResourceOverlay()
+			}
+		}
+	}
+}
+
+func overlayPriceOnInstanceTypes(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType) {
+	// we will only update instance types once
+	overriddenInstanceType := map[string][]string{}
+	for _, overlay := range overlays {
+		// if price or price adjustment is not defined, then we should skip the overlay
+		if overlay.Spec.Price == nil && overlay.Spec.PriceAdjustment == nil {
+			continue
+		}
+		overlaySelector := scheduling.NewRequirements()
+		overlaySelector.Add(scheduling.NewNodeSelectorRequirements(overlay.Spec.Requirements...).Values()...)
+
+		for _, it := range its {
+			if !it.Requirements.IsCompatible(overlaySelector) {
+				continue
+			}
+			if _, ok := overriddenInstanceType[it.Name]; !ok {
+				overriddenInstanceType[it.Name] = []string{}
+			}
+
+			for _, of := range it.Offerings {
+				offeringRequirementsHash := fmt.Sprint(lo.Must(hashstructure.Hash(of.Requirements.String(), hashstructure.FormatV2, &hashstructure.HashOptions{})))
```

Comment:
nit: rather than indexing into the map with an instance type name and the hash, would it be better to just include the it name in the hash and have a single dimensional set?

Reply:
I like that idea and it would be cleaner. Updating the function